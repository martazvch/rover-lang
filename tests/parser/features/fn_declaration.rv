fn add() {}
fn add(arg1: int){}
fn add(arg1:int,) {

}
fn add(
    arg1: int
) {}
fn add(
    arg1: int) {}
fn add(
    arg1: int,
)
{

}

fn add(param1: int, param2: Person,) {}

// Multiline body + no args
fn _ad() -> int {
    var tmp = 5
}

// Multiline parameters + body on new line
fn foo(
    p1: int,
    p2: Creation,
) -> Promise
{}

// Share type
fn shareType(a, b: int, c: Color, d, e, f: bool) {}

// Default value
fn defaultValue(a = 1, b, c = true, d: float, e = "Tom") {}

// Mix types and value
fn mixAll(a, b: int = 4, c = 6., d, e: str, f, g: float = 99) {}
