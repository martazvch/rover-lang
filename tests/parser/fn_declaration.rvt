// Function declaration

code:
fn add() {}
fn add(arg1: int){}
fn add(arg1:int,) {

}
fn add(
    arg1: int
) {}
fn add(
    arg1: int) {}
fn add(
    arg1: int,
)
{

}

fn add(param1: int, param2: Person,) {}

// Multiline body + no args
fn _ad() -> int {
    var tmp = 5
}

// Multiline parameters + body on new line
fn foo(
    p1: int,
    p2: Creation,
) -> Promise
{}

expect:
[Fn declaration add, type void, arity 0
    params:
    body:
        [Block]
]
[Fn declaration add, type void, arity 1
    params:
        arg1, type int
    body:
        [Block]
]
[Fn declaration add, type void, arity 1
    params:
        arg1, type int
    body:
        [Block]
]
[Fn declaration add, type void, arity 1
    params:
        arg1, type int
    body:
        [Block]
]
[Fn declaration add, type void, arity 1
    params:
        arg1, type int
    body:
        [Block]
]
[Fn declaration add, type void, arity 1
    params:
        arg1, type int
    body:
        [Block]
]

[Fn declaration add, type void, arity 2
    params:
        param1, type int
        param2, type Person
    body:
        [Block]
]

[Fn declaration _ad, type int, arity 0
    params:
    body:
        [Block]
            [Var declaration tmp, type void, value
                [Int literal 5]
            ]
]

[Fn declaration foo, type Promise, arity 2
    params:
        p1, type int
        p2, type Creation
    body:
        [Block]
]

================

// Errors

code:
fn
fn add
fn add(
fn add(1)
fn add(a)
fn add(a: int -> bool
fn add() bool
fn add() -> bool 1 + 1
fn a(
    a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,
    a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,
    a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,
    a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,
    a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,
    a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,a:int,
) -> bool {}

error:
ExpectFnName
ExpectParenAfterFnName
ExpectName, parameter
ExpectName, parameter
MissingFnParamType
ExpectParenAfterFnParams
ExpectArrowBeforeFnType
ExpectBraceBeforeFnBody
TooManyFnArgs, parameter

================

