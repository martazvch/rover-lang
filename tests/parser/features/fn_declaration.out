{
    "fn_decl": {
        "name": "add",
        "params": [],
        "return_type": "void",
        "block": []
    },
    "fn_decl": {
        "name": "add",
        "params": [
            {
                "name": "arg1",
                "type": "int"
            }
        ],
        "return_type": "void",
        "block": []
    },
    "fn_decl": {
        "name": "add",
        "params": [
            {
                "name": "arg1",
                "type": "int"
            }
        ],
        "return_type": "void",
        "block": []
    },
    "fn_decl": {
        "name": "add",
        "params": [
            {
                "name": "arg1",
                "type": "int"
            }
        ],
        "return_type": "void",
        "block": []
    },
    "fn_decl": {
        "name": "add",
        "params": [
            {
                "name": "arg1",
                "type": "int"
            }
        ],
        "return_type": "void",
        "block": []
    },
    "fn_decl": {
        "name": "add",
        "params": [
            {
                "name": "arg1",
                "type": "int"
            }
        ],
        "return_type": "void",
        "block": []
    },
    "fn_decl": {
        "name": "add",
        "params": [
            {
                "name": "param1",
                "type": "int"
            },
            {
                "name": "param2",
                "type": "Person"
            }
        ],
        "return_type": "void",
        "block": []
    },
    "fn_decl": {
        "name": "_ad",
        "params": [],
        "return_type": "int",
        "block": [
            "var_decl": {
                "name": "tmp",
                "type": {},
                "value": {
                    "literal": "5"
                }
            }
        ]
    },
    "fn_decl": {
        "name": "foo",
        "params": [
            {
                "name": "p1",
                "type": "int"
            },
            {
                "name": "p2",
                "type": "Creation"
            }
        ],
        "return_type": "Promise",
        "block": []
    }
}
